scanner/matcher/updater/updater.go

```diff
@@ -0,0 +1,326 @@
+		return libvuln.OfflineImport(ctx, u.pool, reader)
+	}
+	return u, nil
+}
+
+func fillOpts(opts *Opts) error {
+	if opts == nil {
+		panic("programmer error")
+	}
+	if err := validate(*opts); err != nil {
+		return err
+	}
+
+	if opts.Client == nil {
+		opts.Client = http.DefaultClient
+	}
+	if opts.Root == "" {
+		opts.Root = os.TempDir()
+	}
+	if opts.UpdateRetention <= 1 {
+		opts.UpdateRetention = libvuln.DefaultUpdateRetention
+	}
+	if opts.UpdateInterval < time.Minute {
+		opts.UpdateInterval = 5 * time.Minute
+	}
+
+	return nil
+}
+
+func validate(opts Opts) error {
+	if opts.Store == nil || opts.Locker == nil || opts.Pool == nil || opts.MetadataStore == nil {
+		return errors.New("must provide a Store, a Locker, a Pool, and a MetadataStore")
+	}
+	if _, err := url.Parse(opts.URL); err != nil {
+		return fmt.Errorf("invalid URL: %q", opts.URL)
+	}
+	return nil
+}
+
+// Start periodically updates the vulnerability data via Update.
+// Each period is adjusted by some amount of jitter.
+func (u *Updater) Start(ctx context.Context) error {
+	ctx = zlog.ContextWithValues(ctx, "component", "matcher/updater/Updater.Start")
+
+	zlog.Info(ctx).Msg("starting initial update")
+	if err := u.update(ctx); err != nil {
+		zlog.Error(ctx).Err(err).Msg("errors encountered during updater run")
+	}
+	zlog.Info(ctx).Msg("completed initial update")
+
+	timer := time.NewTimer(u.updateInterval + jitter())
+	defer timer.Stop()
+	for {
+		select {
+		case <-timer.C:
+			zlog.Info(ctx).Msg("starting update")
+			if err := u.update(ctx); err != nil {
+				zlog.Error(ctx).Err(err).Msg("errors encountered during updater run")
+			}
+			zlog.Info(ctx).Msg("completed update")
+
+			timer.Reset(u.updateInterval + jitter())
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+// update runs a vulnerability data update, once.
+func (u *Updater) update(ctx context.Context) error {
+	ctx = zlog.ContextWithValues(ctx, "component", "matcher/updater/Updater.update")
+
+	ctx, done := u.locker.TryLock(ctx, name)
+	defer done()
+	if err := ctx.Err(); err != nil {
+		zlog.Debug(ctx).
+			Err(err).
+			Str("updater", name).
+			Msg("lock context canceled, excluding from run")
+		zlog.Info(ctx).
+			Str("updater", name).
+			Msg("did not obtain lock, skipping update run")
+		return nil
+	}
+
+	prevTimestamp, err := u.metadataStore.GetLastVulnerabilityUpdate(ctx)
+	if err != nil {
+		zlog.Debug(ctx).
+			Err(err).
+			Msg("did not get previous vuln update timestamp")
+		return err
+	}
+	zlog.Info(ctx).
+		Str("timestamp", prevTimestamp.Format(http.TimeFormat)).
+		Msg("previous vuln update")
+
+	f, timestamp, err := u.fetch(ctx, prevTimestamp)
+	if err != nil {
+		return err
+	}
+	if f == nil {
+		// Nothing to update at this time.
+		zlog.Info(ctx).Msg("no new vulnerability update")
+		return nil
+	}
+	defer func() {
+		if err := f.Close(); err != nil {
+			zlog.Error(ctx).Err(err).Msg("error closing temp update file")
```

Comment:
Nit: The "error" in "error closing..." is redundant. We're emitting an error-level log entry.